BLE Gap deprecation cleanup/rework

LDong-Arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->
Preceding PRs:
* ~#12674~
* ~https://github.com/ARMmbed/mbed-os-example-ble/pull/289 - examples update that can and should get in before this major PR.~ (merged)

In conjunction with
* https://github.com/ARMmbed/mbed-os-example-ble/pull/290 - example update for address types.
### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
* Remove BLE `LegacyGap` class and implementations. Map `ble::Gap` to global `Gap`.
* Add non-deprecated APIs (originally found in `LegacyGap`) to `Gap` and implementations into `GenericGap`.
* Remove/replace legacy BLE event signalling (callback chains) in favour of EventHandler, where possible.
* Remove `BLEProtocol::` and replace `AddressType_t` with `own_address_type_t`/`peer_address_type_t`, `AddressByte_t` with `address_t`, etc.
* Replace `BLE::Address_t` (for accessing address types + values in whitelists) with the new equivalent `whitelist_t::entry_t`.
* Rename `ConnectionParams_t` -> `PreferredConnectionParams_t` and use it only for Generic Access Service.
* Fix misspelt `central_privay_configuration_t` -> `central_privacy_configuration_t`
* Remove legacy type definitions and replace _internal_ usages with new types, e.g. address types, disconnection reason, etc.

Notes:
* `setDeviceName()`, `getDeviceName()`, `setAppearance()`, `getAppearance()` from `LegacyGap` are removed, as device name and appearance can be set via `AdvertisingDataBuilder` and saved by the application. But there were previously no deprecation warnings for them.
* `get/setPreferredConnectionParams()` are dropped from `Gap` - updated versions will be added to Gatt where they should belong to.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
Legacy BLE Gap APIs and related type definitions are no longer available.
Some APIs have been updated to take parameters of new type definitions of address types + values.

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
* Applications based on up-to-date BLE APIs should be able to make most BLE function calls as before. 
* However as mentioned above some deprecated types (e.g. address) were not warned previously - their usages need to be replaced with corresponding new types. The best clue is compilation errors which can be manually fixed. The PR to update mbed-os-example-ble is an example of how to do this: https://github.com/ARMmbed/mbed-os-example-ble/pull/289
* Applications should set device name and appearance using `AdvertisingDataBuilder` and internally remember their values as needed, since APIs to get/set them are removed as explained above.

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
API documentation is generated with Doxygen. Examples quoted in documentation (e.g. [this](https://os.mbed.com/docs/mbed-os/v5.15/apis/ble.html)) are based on ARMmbed/mbed-os-example-ble which requires an update.

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [x] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR

Manual testing
* BLE examples compile & work as before using https://github.com/ARMmbed/mbed-os-example-ble/pull/290
* `ble-cliapp` is updated to work with this PR and used to manually test APIs (by referencing Gap tests in `ble-tests`)

----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

LDong-Arm committed:
ble/Gap.h: remove deprecated APIs

LDong-Arm committed:
Remove deprecated implementations from GenericGap & LegacyGap

LDong-Arm committed:
GenericGap: remove legacy dis/connection callbacks signalling

LDong-Arm committed:
BLE Gap: remove version one scan API

LDong-Arm committed:
GenericGap: prepare legacy advertising set

This is based on (and supercedes) PR #10772.

As the switches between old and new APIs do not exist anymore,
we create a function for legacy advertising set creation.

LDong-Arm committed:
GenericGap: remove legacy timeout callback signalling

LDong-Arm committed:
Move BLE whitelist APIs from LegacyGap to ble::Gap

LDong-Arm committed:
BLE: update to use whitelist_t from ble namespace

LDong-Arm committed:
Cordio: remove LegacyGap implementation

LDong-Arm committed:
BLE: Remove LegacyGap and related types

LDong-Arm committed:
BLE: construct disconnection_reason_t from uint8_t (received from HCI)

LDong-Arm committed:
Add to Gap/GenericGap non-deprecated APIs originally in LegacyGap

LDong-Arm committed:
BLE: use up-to-date structs for disconnections

LDong-Arm committed:
GenericAccessService: remove unused device name/appearance functions due to API changes

LDong-Arm committed:
GenericGap: clean up unused helper functions

LDong-Arm commented on features/FEATURE_BLE/ble/Gap.h:
```diff
@@ -17,2687 +17,15 @@
-// import LegacyGap implementation into global namespace
-typedef ble::impl::LegacyGap Gap;
+// import Gap implementation into global namespace
+typedef ble::impl::Gap Gap;
```
(Comment to force rendering large diff...)

LDong-Arm replied:
still not rendered...

LDong-Arm commented on features/FEATURE_BLE/source/generic/GenericGap.tpp:
```diff
@@ -28,8 +28,6 @@
 
-#include "ble/pal/Deprecated.h"
-
 namespace ble {
```
(Comment to force rendering large diff...)

LDong-Arm commented on features/FEATURE_BLE/ble/generic/GenericGap.h:
```diff
@@ -28,7 +28,6 @@
 #include "ble/pal/GenericAccessService.h"
 #include "ble/pal/EventQueue.h"
 #include "ble/pal/ConnectionEventMonitor.h"
-#include "ble/pal/Deprecated.h"
```
(Comment to force rendering large diff...)

LDong-Arm committed:
GenericGap: ensure legacy advertising set exists when setting data

LDong-Arm committed:
BLE LinkLossService: use new EventHandler API to handle disconnection

LDong-Arm committed:
BLE: remove ResolutionStrategy type

LDong-Arm committed:
Add whitelist_t::entry_t which will replace Address_t to store addr type and value

LDong-Arm committed:
BLE: replace legacy address types (BLEProtocol) with new ones

LDong-Arm committed:
Remove BLEProtocol.h

LDong-Arm committed:
Add default initialisation of own_address_type_t

LDong-Arm commented on features/FEATURE_BLE/ble/BLETypes.h:
```diff
@@ -831,6 +831,70 @@ struct coded_symbol_per_bit_t :SafeEnum<coded_symbol_per_bit_t, uint8_t> {
+     * It contains an address-type (peer_address_type_t) and the address value
+     * (address_t).
+     */
+    struct entry_t {
```
Note: replaces `BLEProtocol::Address_t`

LDong-Arm commented on features/FEATURE_BLE/ble/gap/Gap.h:
```diff
@@ -279,6 +279,21 @@ class Gap : public StaticInterface<Impl, Gap> {
+     *
+     * @see Gap::onShutdown().
+     */
+    typedef FunctionPointerWithContext<const Gap *> GapShutdownCallback_t;
```
Note: This is not in the EventHandler, to align with the bahaviour of Gatt, SecurityManager, etc.
(And the resetting of Gap causes EventHandler pointer to be reset)

LDong-Arm commented on features/FEATURE_BLE/ble/gap/Gap.h:
```diff
@@ -532,6 +547,53 @@ class Gap : public StaticInterface<Impl, Gap> {
+         * @note maxConnectionInterval is in ms in the formulae above.
+         */
+        uint16_t connectionSupervisionTimeout;
+    } ConnectionParams_t;
```
Need to remove this - it looks like there's still some clean up to do

LDong-Arm commented on features/FEATURE_BLE/source/generic/GenericGap.tpp:
```diff
@@ -284,64 +212,47 @@ static bool is_random_private_resolvable_address(
+                }
+                break;
+            default: // ANONYMOUS
                 return false;
```
It's kind of reworked - the old and new address types do not match exactly...

pan- replied:
That's fair

pan- commented on features/FEATURE_BLE/ble/gap/Types.h:
```diff
@@ -854,6 +860,17 @@ struct disconnection_reason_t : SafeEnum<disconnection_reason_t, uint8_t> {
+     *
+     * @note This should only be used for casting raw values from HCI.
+     */
+    disconnection_reason_t(uint8_t value) : SafeEnum(value)
```
Could you make the constructor explicit ? Otherwise that defeats the purpose of having a safe enum.

pan- commented on features/FEATURE_BLE/source/generic/GenericGap.tpp:
```diff
@@ -284,64 +212,47 @@ static bool is_random_private_resolvable_address(
+                }
+                break;
+            case peer_address_type_t::RANDOM:
+                if (!is_random_static_address(entry.address)
```
The condition can be replace by a test to `is_random_private_non_resolvable_address`.

LDong-Arm replied:
Checking `is_random_private_non_resolvable_address` alone doesn't guard against the scenario where the address value isn't even random.

It brings in another question: the old code only allowed public and random static, whereas the documentation of setWhitelist suggests resolvable private should also work. Were there any limitation that prevented us from whitelisting resolvable private addresses?

pan- replied:
There's not much reason of adding a resolvable private address to the list as it change very regularly, you want to put the _real_ address obtained during pairing in the whitelist, the controller should ensure derived resolvable addresses are check against this address.

LDong-Arm replied:
I'll change it. With real addresses things would be simpler here.

LDong-Arm replied:
done

pan- commented on features/FEATURE_BLE/source/generic/GenericGap.tpp:
```diff
@@ -1656,18 +966,6 @@ void GenericGap<PalGapImpl, PalSecurityManager, ConnectionEventMonitorEventHandl
-template <template<class> class PalGapImpl, class PalSecurityManager, class ConnectionEventMonitorEventHandler>
-void GenericGap<PalGapImpl, PalSecurityManager, ConnectionEventMonitorEventHandler>::process_scan_timeout()
-{
-    ble_error_t err = _pal_gap.scan_enable(false, false);
```
Could you keep the part of the code that disable scanning ? That is useful for BT4 devices where this feature isn't part of the controller.

LDong-Arm replied:
This function was unused even before this PR. The disabling of scan for BT4 is done in `process_legacy_scan_timeout()` above.

pan- replied:
I missed it thanks, indeed this is not needed.

pan- commented on features/FEATURE_BLE/source/generic/GenericGap.tpp:
```diff
@@ -2509,6 +1786,8 @@ ble_error_t GenericGap<PalGapImpl, PalSecurityManager, ConnectionEventMonitorEve
         return BLE_ERROR_INVALID_PARAM;
     }
 
+    prepare_legacy_advertising_set();
```
Is it necessary to call that function if the advertising handle is not the legacy one ?

LDong-Arm replied:
I'll add a check for legacy handle

pan- commented on features/FEATURE_BLE/targets/TARGET_CORDIO/source/CordioGattServer.cpp:
```diff
@@ -818,47 +818,6 @@ void GattServer::setPreferredConnectionParams(const ::Gap::ConnectionParams_t& p
     memcpy(generic_access_service.ppcp + 6, &params.connectionSupervisionTimeout, 2);
 }
 
-ble_error_t GattServer::setDeviceName(const uint8_t *deviceName)
```
I would keep these functions inside the cordio gatt server, they will be needed once the feature is exposed publicly.

pan- commented on features/FEATURE_BLE/targets/TARGET_CORDIO/CordioPalGenericAccessService.h:
```diff
@@ -19,69 +19,6 @@ class GenericAccessService : public ::ble::pal::GenericAccessService {
 
     virtual ~GenericAccessService() { }
 
-    virtual ble_error_t get_device_name_length(uint8_t& length) {
```
Can we keep these functions in Cordio, it will be easier for us if they are here to enable publicly the feature.

LDong-Arm replied:
Those functions use some deprecated types (e.g. `GapAdvertisingData` which is replaced by `AdvertisingDataBuilder`). Can we use macros (e.g. `#if 0`) to disable them, and fix them once we reenable the features?

pan- replied:
that's ok if remove the code with `#if 0`. Please make sure to have a ticket open for these.

pan- reviewed (changes requested):
Thanks Lingkai, it is very promising. I've noticed few improvements that could be applied here and there.

LDong-Arm reviewed (commented):
@pan- Thanks for the review, I'm working on addressing your comments.

LDong-Arm committed:
disconnection_reason_t from raw uint8_t should be explicited an undocumented

LDong-Arm committed:
Gap: create legacy advertising set only when required

LDong-Arm committed:
Cordio: bring back device name & appearance function

Functions in Cordio for device name & appearance are currently
unused as they are both part of and dependent on deprecated
Mbed OS BLE APIs.

Nonetheless we want to keep them (and disable using macros)
so we can reintroduce them and make improvements in the future
as needed.

LDong-Arm committed:
Remove connection_params from internal on_connected()

The callback on_connected() in ConnectionEventMonitorEventHandler
contains a parameter ConnectionParams_t which is superceded by
ConnectionParameters. Since it's not used, remove it.

LDong-Arm committed:
BLE: correct misspelling: prefered -> preferred

LDong-Arm committed:
Rename ConnectionParams_t -> PreferredConnectionParams_t

For actual connections, full ConnectionParameters is used. But
as per BLE specification, Generic Access Service can display
preferred connection parameters which is a smaller subset and
ConnectionParams_t matches exactly. Thus we rename/repurpose
it to PreferredConnectionParams_t.

LDong-Arm committed:
Gap.tpp: remove misleading deprecation header

LDong-Arm committed:
Remove a few leftovers in BLE namespace

pan- reviewed (approved):
Thanks @LDong-Arm that looks very good, did you had the change to try it with ble-cliapp and the test suite ?

LDong-Arm committed:
BLE whitelist can only contain public or random static addresses

pan- reviewed (approved):
It looks good from all perspectives, thanks @LDong-Arm for your efforts on the test app and test suite!
This PR can be merged when CI pass.
